### ISSUE TITLE
[Reliability] Negative pending job count in batches after enabling super fetch

### ISSUE THREAD

**Issue by seckenrode** [mperham/sidekiq#3646]
Ruby version: ruby 2.3.4p301 (2017-03-30 revision 58214) [x86_64-linux]
Sidekiq / Pro / Enterprise version(s): Sidekiq v4.2.10/Sidekiq Pro v3.5.0

Hello!

We've been experimenting with enabling super_fetch on our heroku-deployed sidekiq workers to get better reliability against operational issues.  Most of the issues we've encountered have been due to misbehaviors on our part and we've been addressing them.  We are, however, seeing batches end up in negative pending status due to jobs completing multiple times.

Based on #3019 and #3130, this appears to be a known situation, but I'm interested in learning a bit more about how we can either catch or debug where our behavior could be wrong.

1. These batches do not enqueue jobs after the initial job set, and they use `push_bulk` to ensure that we send all batches at the same time.
2. This typically happens on higher volume, but even small batches have this issue. [1]
3. Our biggest culprits for problems with reliability features so far have been either pathologically bad behavior from our jobs (consistently using enough memory that heroku SIGKILLs the process or heartbeats stop going through), but these cases do not seem to happen every time we get negative pending count.

Do you have any advice for how to address this or is there additional information we can provide that might help find a root cause?

Thanks,
Steve

[1] <img width="1188" alt="screen shot 2017-10-25 at 8 32 51 am" src="https://user-images.githubusercontent.com/3323794/31999025-d83dd4cc-b95f-11e7-9de0-ce00b01c8bb8.png">

**Comment by aaroalan** [comment]
I got this issue and did not get the "real" reason why it was happening, but the solution in my project was execute the callback job in a different queue than the jobs in the batch. I used a priority queue to run the callback immediately.

**Comment by mperham** [comment]
In my experience, negative batch counts are 100% caused by improper shutdown of a worker process while jobs are in progress.  You need to make sure all of your processes, including any autoscaled ones, are cleanly shut down within Heroku's shutdown timeout (30 sec).  Leverage the quiet signal as much as possible to maximize your shutdown window, as explained on the Deployment wiki page.

**Comment by seckenrode** [comment]
Thanks so much @mperham! I'm digging in a bit more to this and am finding some strange behavior that you might have insight on.

It turns out we have a very small timeout value set (2 seconds) that was intended to hint our workers to just push any pending work back to redis immediately after getting a TERM signal from heroku.  From the looks of it, however, it appears that there might be some strange behavior going on.

Here's a log snippet

```
...
Dec 13 12:49:11 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6l8w Panorama::Job JID-d573912a8bb2488afc638bd1 BID-xVM2eiLyzb1XoQ INFO: start 
Dec 13 12:49:13 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6ll4 Panorama::Job JID-6020576f4023b9a0a3743b71 BID-xVM2eiLyzb1XoQ INFO: start 
Dec 13 12:49:13 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:49:13.453 28 LOG Got SIGTERM, ignoring! 
Dec 13 12:49:13 panorama-rainbow app/low_power_worker.5:  27 TID-otnkhv95c INFO: Shutting down 
Dec 13 12:49:13 panorama-rainbow app/low_power_worker.5:  27 TID-otnkhv95c INFO: Terminating quiet workers 
Dec 13 12:49:14 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6lz0 INFO: Scheduler exiting... 
Dec 13 12:49:14 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6l8w Panorama::Job JID-d573912a8bb2488afc638bd1 BID-xVM2eiLyzb1XoQ INFO: done: 3.457 sec 
Dec 13 12:49:14 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6lr8 Panorama::Job JID-62c669745aaf5549f3b8b320 BID-xVM2eiLyzb1XoQ INFO: done: 3.745 sec 
Dec 13 12:49:14 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6lf0 Panorama::Job JID-2a58dff62d6919706da03916 BID-xVM2eiLyzb1XoQ INFO: done: 3.744 sec 
Dec 13 12:49:15 panorama-rainbow app/low_power_worker.5:  27 TID-otnkhv95c INFO: Pausing to allow workers to finish... 
Dec 13 12:49:15 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6l2s Panorama::Job JID-e8b13dad7b8ebdb4545e4b73 BID-xVM2eiLyzb1XoQ INFO: done: 3.361 sec 
Dec 13 12:49:15 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6ll4 Panorama::Job JID-6020576f4023b9a0a3743b71 BID-xVM2eiLyzb1XoQ INFO: done: 2.656 sec 
Dec 13 12:49:15 panorama-rainbow app/low_power_worker.5:  27 TID-otnkhv95c WARN: Terminating 1 busy worker threads 
Dec 13 12:49:15 panorama-rainbow app/low_power_worker.5:  27 TID-otnkhv95c WARN: Work still in progress [#<struct Sidekiq::Pro::SuperFetch::Retriever::UnitOfWork queue="queue:low_power_normal_priority", job="{\"class\":\"Panorama::Job\",\"args\":[...],\"queue\":\"low_power_normal_priority\",\"retry\":true,\"jid\":\"6020576f4023b9a0a3743b71\",\"created_at\":1513187352.0000834,\"bid\":\"xVM2eiLyzb1XoQ\",\"enqueued_at\":1513187352.0003126}", local_queue="queue:sq|low_power_worker.5:27:fcbff4ab8d3c|low_power_normal_priority">] 
Dec 13 12:49:15 panorama-rainbow app/low_power_worker.5:  27 TID-otnkhv95c INFO: Bye! 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:49:15.915 28 LOG C-0x1600458: db1/u6cnvv3ace71i7@127.0.0.1:35230 closing because: client close request (age=121) 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:49:15.916 28 LOG C-0x1600768: db1/u6cnvv3ace71i7@127.0.0.1:35239 closing because: client close request (age=121) 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:49:15.916 28 LOG C-0x16002d0: db1/u6cnvv3ace71i7@127.0.0.1:34563 closing because: client close request (age=132) 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:49:15.919 28 LOG C-0x16008f0: db1/u6cnvv3ace71i7@127.0.0.1:35252 closing because: client close request (age=121) 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:49:15.919 28 LOG C-0x16005e0: db1/u6cnvv3ace71i7@127.0.0.1:35235 closing because: client close request (age=121) 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:49:15.919 28 LOG C-0x1600a78: db1/u6cnvv3ace71i7@127.0.0.1:35258 closing because: client close request (age=121) 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=app-end 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=exit process=app 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=kill-app pid=26 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  bin/start-pgbouncer-stunnel: line 51: kill: (-26) - No such process 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=wait-app pid=26 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=kill-aux name=pgbouncer pid=24 signal=SIGINT 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=kill-aux name=stunnel pid=23 signal=SIGINT 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=stunnel-end 
Dec 13 12:49:16 panorama-rainbow app/low_power_worker.5:  27 TID-otnma6l2s Panorama::Job JID-e8b13dad7b8ebdb4545e4b73 BID-xVM2eiLyzb1XoQ INFO: start 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=pgbouncer-enabled 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=config-gen-start 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  Setting DATABASE_URL_PGBOUNCER config var 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=config-gen-end 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=config-gen-override DATABASE_URL 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=stunnel-launched pid=23 signal=SIGINT 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=stunnel-start 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=pgbouncer-launched pid=25 signal=SIGINT 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=app-launched pid=26 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=app-start 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  buildpack=pgbouncer at=pgbouncer-start 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:51:00.754 29 LOG File descriptor limit: 10000 (H:10000), max_client_conn: 100, max fds possible: 114 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:51:00.755 29 LOG listening on 127.0.0.1:6000 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:51:00.755 29 LOG listening on unix:/tmp/.s.PGSQL.6000 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:51:00.755 29 LOG process up: pgbouncer 1.7, libevent 2.0.21-stable (epoll), adns: evdns2 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG5[ui]: stunnel 5.28 on x86_64-unknown-linux-gnu platform 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG5[ui]: Compiled/running with OpenSSL 1.0.1f 6 Jan 2014 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG5[ui]: Threading:PTHREAD Sockets:POLL,IPv6 TLS:ENGINE,FIPS,OCSP,PSK,SNI 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG5[ui]: Reading configuration from file vendor/stunnel/stunnel-pgbouncer.conf 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG5[ui]: UTF-8 byte order mark not detected 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG5[ui]: FIPS mode disabled 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG4[ui]: Service [DATABASE_URL] needs authentication to prevent MITM attacks 
Dec 13 12:51:01 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:00 LOG5[ui]: Configuration successful 
Dec 13 12:51:06 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:51:06.507 29 LOG C-0x160d2d0: db1/u6cnvv3ace71i7@127.0.0.1:43951 login attempt: db=db1 user=u6cnvv3ace71i7 tls=no 
Dec 13 12:51:06 panorama-rainbow app/low_power_worker.5:  2017-12-13 17:51:06.507 29 LOG S-0x1611f80: db1/u6cnvv3ace71i7@unix:6101 new connection to server 
Dec 13 12:51:06 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:06 LOG5[0]: Service [DATABASE_URL] accepted connection from unnamed socket 
Dec 13 12:51:06 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:06 LOG5[0]: s_connect: connected 34.200.211.67:5432 
Dec 13 12:51:06 panorama-rainbow app/low_power_worker.5:  2017.12.13 17:51:06 LOG5[0]: Service [DATABASE_URL] connected remote server from 172.19.64.126:34001 
Dec 13 12:51:06 panorama-rainbow app/low_power_worker.5:  27 TID-ovt4pnhfw INFO: Booting Sidekiq 4.2.10 with redis options {:url=>"redis://rediscloud:REDACTED@redis-16075.c8.us-east-1-4.ec2.cloud.redislabs.com:16075", :namespace=>"panorama-rainbow-sidekiq", :driver=>:hiredis} 
Dec 13 12:51:15 panorama-rainbow app/low_power_worker.5:  27 TID-ovt4pnhfw INFO: Running in ruby 2.3.4p301 (2017-03-30 revision 58214) [x86_64-linux] 
Dec 13 12:51:15 panorama-rainbow app/low_power_worker.5:  27 TID-ovt4pnhfw INFO: Sidekiq Pro 3.5.0, commercially licensed.  Thanks for your support! 
Dec 13 12:51:15 panorama-rainbow app/low_power_worker.5:  27 TID-ovt4pnhfw INFO: Starting processing, hit Ctrl-C to stop 
Dec 13 12:51:15 panorama-rainbow app/low_power_worker.5:  27 TID-ovt5rkmuw INFO: SuperFetch activated 
Dec 13 12:51:15 panorama-rainbow app/low_power_worker.5:  27 TID-ovt5rhzmk Panorama::Job JID-d3a0956002135c2273e36ddb BID-xVM2eiLyzb1XoQ INFO: start 
...
```

What I'm most curious about is that it appears that after the TERM signal is received, and after the `Shutting down` and `Terminating quiet workers` messages, we go through the cleanup, sidekiq pushes a single WIP job back into the SuperFetch queue, we get the `Bye!` message, then a second later JID e8b13dad7b8ebdb4545e4b73 is picked up for work by this worker.

This same JID was marked as done between the time the `Pausing Workers` message was received and the `Bye!` message was received, but is at the same second timestamp.

I have adjusted our timeout to be longer (20 seconds) to conform more with the conventions and recommendations in https://github.com/mperham/sidekiq/wiki/Signals#term, but I am interested in understanding how that completed job might have been picked up for work *after* the process was quieted and *after* it was marked as done by sidekiq.

Thanks!

P.S. Please let me know if I should elevate this as another issue, I am happy to!

**Comment by mperham** [comment]
There is a race condition between batch job success and Sidekiq removing the job as work in progress.  The batch job finishes but at the same time, Sidekiq raises Sidekiq::Shutdown to terminate any work in progress and push the job back to Redis.  This is more likely with a shorter timeout.  Extending the timeout will minimize the possibility; unfortunately fixing this race would require a substantial redesign of Sidekiq internals.

**Comment by seckenrode** [comment]
Good to know!  We've increased the close timeout here and hopefully that fixes things up a bit more reliably.  I'll report back if we still are encountering weirdness.

**Comment by joekur** [comment]
> Extending the timeout will minimize the possibility; unfortunately fixing this race would require a substantial redesign of Sidekiq internals.

@mperham was this race condition fixed in 3.7.0 or do you believe it still exists? I'm trying to debug this same issue in our app. We're on sidekiq-pro 3.7.1 and using super_fetch.

**Comment by mperham** [comment]
The race condition was dramatically lessened in 3.7. I’ve seen no new negative counts issues since that version was released.
